shared/src/main/scala/fetch.scala

```diff
@@ -17,207 +17,467 @@
 package fetch
 
 import scala.collection.immutable.Map
+import scala.util.control.NoStackTrace
 
-import cats.{Applicative, Eval}
-import cats.data.{NonEmptyList, StateT}
-import cats.free.Free
+import scala.concurrent.duration.MILLISECONDS
+
+import cats._
 import cats.instances.list._
+import cats.effect._
+import cats.effect.concurrent.{Ref, Deferred}
+import cats.syntax.all._
+import cats.data.NonEmptyList
 
-trait FetchException extends Throwable with Product with Serializable {
-  def env: Env
-}
-case class NotFound(env: Env, request: FetchOne[_, _]) extends FetchException
-case class MissingIdentities(env: Env, missing: Map[DataSourceName, List[Any]])
-    extends FetchException
-case class UnhandledException(env: Env, err: Throwable) extends FetchException
 
-/** Requests in Fetch Free monad. */
-sealed trait FetchRequest extends Product with Serializable
+object `package` {
+  // Fetch queries
+  sealed trait FetchRequest extends Product with Serializable
 
-sealed trait FetchQuery[I, A] extends FetchRequest {
-  def dataSource: DataSource[I, A]
-  def identities: NonEmptyList[I]
-}
+  // A query to a remote data source
+  sealed trait FetchQuery extends FetchRequest {
+    def dataSource[I, A]: DataSource[I, A]
+    def identities[I]: NonEmptyList[I]
+  }
+  case class FetchOne[I, A](id: I, ds: DataSource[I, A]) extends FetchQuery {
```

Comment:
This definition seems dangerous. It says that the underlying datasource has fixed types for `I` and `A`, but then the definitions below allow callers to specify the types.